File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    if target not in words:
        return 0  # target이 words에 없는 경우 변환할 수 없으므로 0 반환

    answer = 0
    visited = [0] * len(words)
    q = deque()
    q.append((begin, 0))
    
    while q:
        word, cnt = q.popleft()
        if word == target:
            answer = cnt
            break
        for i in range(len(words)):
            if not visited[i]:
                temp_cnt = 0
                for j in range(len(word)):
                    if words[i][j] != word[j]:
                        temp_cnt += 1
                if temp_cnt == 1:  # 한 글자만 다른 경우에만 큐에 추가
                    q.append((words[i], cnt + 1))
                    visited[i] = 1
    
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    static class Pair{
        String word;
        int level;
        
        public Pair(String word, int level) {
            this.word = word;
            this.level = level;
        }
    }
    static boolean[] visited;
    public int solution(String begin, String target, String[] words) {
        visited = new boolean[words.length];
        int answer = bfs(begin, target, words);
        return answer;
    }
    
    public int bfs(String begin, String target, String[] words) {
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(begin, 0));
        
        while (!q.isEmpty()) {
            Pair curr = q.poll();
            String word = curr.word;
            int level = curr.level;
            
            if (word.equals(target)) {
                return level;
            }
            
            for (int i = 0; i<words.length; i++) {
                if (!visited[i]) {
                    int dist = 0;
                    
                    for (int j = 0; j<word.length(); j++) {
                        if (words[i].charAt(j) != word.charAt(j)) dist++;
                    }
                    if (dist==1){
                        visited[i] = true;
                        q.add(new Pair(words[i], level+1));
                    }
                }
            }
        }
        return 0;
    }
}